### ISSUE TITLE
Graphics drawPolygon error when exceeding PIXI.GraphicsGeometry.BATCHABLE_SIZE

### ISSUE THREAD

**Issue by jjxtra** [pixijs/pixi.js#6047]
### Current Behavior
When using drawPolygon or moveTo and lineTo and exceeding PIXI.GraphicsGeometry.BATCHABLE_SIZE, lines and fills become random and corrupted.

Graphics.drawPolygon also does not seem to work with large arrays in this example, although I have had success with drawPolygon with this same sample data set in my own javascript code offline.

### Possible Solution
Settings PIXI.GraphicsGeometry.BATCHABLE_SIZE = 2147483647; is a work-around, seems to fix the issue. Removing the begin and end fill removes most of the corruption, but a good portion of the lines do not render. Again, setting PIXI.GraphicsGeometry.BATCHABLE_SIZE = 2147483647 fixes the issue.

### Steps to Reproduce
I've created a playground, see link at bottom here.

### Environment
https://cdnjs.cloudflare.com/ajax/libs/pixi.js/5.1.2/pixi.min.js
Chrome, firefox, edge browsers tested on Windows 10 x64.

### Notes
My guess is that the batch size is doing something bad like splitting polygons mid-list or something. Maybe the batch size could be a soft limit and could be changed to not split polygons like that (if indeed that is the problem).

It seems like the problem is in a non-batchable buildDrawCalls path, I *think* that is where the bug is.

If anyone has any ideas why my drawPolygon in the playground does not work but the moveTo / lineTo does, I'd love to know.

### Example code
This playground shows the problem. Uncomment the third line to fix everything.

https://www.pixiplayground.com/#/edit/GeOcBoBtOW2FC005U8cLF

**Comment by ivanpopelyshev** [comment]
ITs a known issue,. There's a workaround

https://github.com/pixijs/pixi.js/issues/5973

**Comment by eXponenta** [comment]
polygon is very big, maximal length - 65k, because used 16 bit index buffer.

**Comment by ivanpopelyshev** [comment]
Not much, but i just dont have time to submit PR because I also have to think more about consequences, something tells me that it can fail in some cases. 

https://github.com/pixijs/pixi.js/blob/dev/packages/graphics/src/GraphicsGeometry.js#L573

```js
if (attrib >= 65536 && settings.HAS_UINT32_INDEX) {
	// TODO: make sure its not batchable!
	this.indicesUint16 = new Uint32Array(this.indices) as any;
} else {
	this.indicesUint16 = new Uint16Array(this.indices);
}
```

`settings.HAS_UINT32_INDEX` is `false` by default, but its modified in `ContextSystem` :

For webgl1: 

```js
settings.HAS_UINT32_INDEX = !!this.extensions.uint32ElementIndex;
```

For webgl2: 

```js
settings.HAS_UINT32_INDEX = true;
```

**Comment by jjxtra** [comment]
Looks pretty innocent. Maybe the variable name should just be indices :)

**Comment by aaclayton** [comment]
Is there a path forward on this? Seems to be coming up a number of times in duplicate tickets. I feel like it's not really that hard to stumble into the problem, for example in my use case:

![image](https://user-images.githubusercontent.com/2468703/68004953-b19d0c00-fc30-11e9-8057-0e0ceeb3406a.png)

I'm drawing many polygons to a Graphics object to create a hexagonal grid. The dimensions of the map do not need to be that large to reach the buffer limit, just a 100 by 100 grid produces 60,000 lines (assuming that's how `drawPolygon` works).

It seems like it's viewed that this is an abnormal edge case that shouldn't occur under normal circumstances, but it doesn't seem to me to be that difficult for a user to encounter this issue without knowing the secret workaround.

**Comment by aaclayton** [comment]
This issue (and many others related to it) have been closed either automatically or manually without - as far as I can tell - a path foward or long-term resolution suggested. 

I am aware of the workaround suggested in the "v5 Hacks" page here (https://github.com/pixijs/pixi.js/wiki/v5-Hacks#removing-65k-vertices-limitation), but as stated in that workaround 

> Its dangerous zone, because some devices browsers dosn't support webgl2

There are many issues related to this behavior:

- https://github.com/pixijs/pixi.js/issues/6184
- https://github.com/pixijs/pixi.js/issues/6107
- https://github.com/pixijs/pixi.js/issues/6104
- https://github.com/pixijs/pixi.js/issues/6119

I'm continuing to maintain the "workaround" with a TODO tag in my code - but I'd like to encourage an official statement about what the correct usage should be.

Suppose you need to render 10,000 polygons. I think it's not obvious - especially to a new user - that you should not simply instantiate one `PIXI.Graphics` instance and iterativerly call `g.drawPolygon(...)` for each required piece of data.

In fact, I would think that perhaps this is the type of behavior that I would generally expect a library like PIXI to figure out for me "under the hood". It's a bit strange to have a method like `drawLine(...)` but you can only call it some number of times before you should instead create a second instance to keep calling the same method.

Anyways - I am willing to change my approach, I have delayed in doing so until learning what the official guidance is. I think it would be good to re-open this to achieve a more formal resolution.

**Comment by jjxtra** [comment]
I agree 100%. This is a deficiency in the pixi library. It will likely take a pull request from someone who really knows the library to get it changed.

**Comment by bigtimebuddy** [comment]
Because this issue keep resurfacing for people, we should try to improve the ergonomics of this use-case. The wiki hacks are insufficient. @ivanpopelyshev any ideas?

**Comment by eXponenta** [comment]
Some engine has limitations on 65k vert per renderable objects, and some browsers can't support Uint32 for webgl1 ( hello IE11).
But you should have knowledge about rendering process when u try render large object.

This aren't basic case.


<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by bigtimebuddy** [comment]
I understand there are limitations, but I think users should be able to more easily opt-in to unit32/webgl2 if they desire. Basically, lets harden the hack into a legit and useful API:

```js
const graphics = new Graphics();
graphics.allow32 = true; // should throw error for environments without OES_element_index_uint
// ... do stuff
```

**Comment by eXponenta** [comment]
There are differences between Graphics and GraphicsGeometry, u always can pass instance of special version of GraphicsGeometry to constructor:) 
Official API:) 
But ok, we can check settings flag internaly, I will make pr for this....

**Comment by eXponenta** [comment]
But this is will be hard PR, because batcher use uint16  too:)

**Comment by bigtimebuddy** [comment]
Users should not be changing BATCHABLE_SIZE to push through some enormous geometry to get batched with sprites, meshes, etc. That's not what that was intended for. Large geometries weren't intended to go through the batcher.

**Comment by eXponenta** [comment]
Default BATCHABLE_SIZE is very small. 
I changed it to 100 000 without performance drop by batch population, and it faster that render 1000 polygons with 100 vertex per element. 
Of course batching to very large mesh not needing.

**Comment by bigtimebuddy** [comment]
Fixed in #6417
